src/daemon/rpc_command_executor.cpp

```diff
@@ -1991,6 +1991,16 @@ bool rpc_command_executor::prepare_registration()
     tools::fail_msg_writer() << "Unable to prepare registration: this daemon is not running in --service-node mode";
     return false;
   }
+  else if (res.last_lokinet_ping == 0 && !force_registration)
+  {
+    tools::fail_msg_writer() << "Unable to prepare registration: this daemon has not received a ping from lokinet yet";
+    return false;
+  }
+  else if (res.last_storage_server_ping == 0 && !force_registration)
+  {
+    tools::fail_msg_writer() << "Unable to prepare registration: this daemon has not received a ping from the storage server yet";
+    return false;
+  }
```

Comment:
I think this can be made a little tighter; oxend normally gets pings from lokinet & ss every 30 seconds now, so we ought to be able to usefully do: `res.last_lokinet_ping < time(nullptr) - 60`

Reply:
There's also a `get_human_time_ago` function in here so you could do:

```C++
tools::fail_msg_writer() << "Unable to prepare registration: this daemon has not received a ping from lokinet "
    << (res.last_lokinet_ping == 0 ? "yet" : "since " + get_human_time_ago(res.last_lokinet_ping, std::time(nullptr)));
```
which should give a message such as:
"... not received a ping from lokinet since 37.2 minutes ago"
when we have received one, but not recently.